VerifyProfileEmail/handler.ts

```diff
@@ -0,0 +1,232 @@
+
+// A token must be in the following format:
+// ULID + ":" + crypto.randomBytes(12)
+const TokenQueryParam = PatternString("^[A-Za-z0-9]{26}:[A-Fa-f0-9]{24}$");
+type TokenQueryParam = t.TypeOf<typeof TokenQueryParam>;
+
+type IVerifyProfileEmailHandler = (
+  context: Context,
+  token: string
+) => Promise<IResponsePermanentRedirect | IResponseErrorValidation>;
+
+/**
+ * Returns a ValidUrl in case of success
+ */
+function verificationSuccessUrl(verificationCallbackUrl: string): ValidUrl {
+  return { href: `${verificationCallbackUrl}?result=success` };
+}
+
+/**
+ * Returns a ValidUrl with a error in case of failure
+ */
+function verificationFailureUrl(
+  verificationCallbackUrl: string,
+  error: string
+): ValidUrl {
+  return { href: `${verificationCallbackUrl}?result=failure&error=${error}` };
+}
+
+const TokenQueryParamMiddleware = RequiredQueryParamMiddleware(
+  "token",
+  TokenQueryParam
+);
+
+// Used in the callback
+export enum VerificationErrors {
+  GENERIC_ERROR = "GENERIC_ERROR",
+  INVALID_TOKEN = "INVALID_TOKEN",
+  TOKEN_EXPIRED = "TOKEN_EXPIRED"
+}
+
+export function VerifyProfileEmailHandler(
+  tableService: TableService,
+  verificationTokensTableName: string,
+  profileModel: ProfileModel,
+  verificationCallbackUrl: string
+): IVerifyProfileEmailHandler {
+  return async (context, token) => {
+    const logPrefix = `VerifyProfileEmail`;
+
+    // STEP 1: Find and verify verification token
+
+    // Get required data
+    const [tokenId, validator] = token.split(":");
+    const validatorHash = crypto
+      .createHash("sha256")
+      .update(validator)
+      .digest("hex");
+
+    // Retrive the entity from the table storage
+    const errorOrMaybeVerificationTokenEntity = await retrieveTableEntity(
+      VerificationTokenEntity,
+      tableService,
+      verificationTokensTableName,
+      tokenId,
+      validatorHash
+    );
+
+    if (isLeft(errorOrMaybeVerificationTokenEntity)) {
+      context.log.error(
+        `${logPrefix}|Error searching verification token|ERROR=${errorOrMaybeVerificationTokenEntity.value}`
+      );
+      return ResponsePermanentRedirect(
+        verificationFailureUrl(
+          verificationCallbackUrl,
+          VerificationErrors.GENERIC_ERROR
+        )
+      );
+    }
+
+    const maybeVerificationTokenEntity =
+      errorOrMaybeVerificationTokenEntity.value;
+
+    if (isNone(maybeVerificationTokenEntity)) {
+      context.log.error(`${logPrefix}|Verification token not found`);
+      return ResponsePermanentRedirect(
+        verificationFailureUrl(
+          verificationCallbackUrl,
+          VerificationErrors.INVALID_TOKEN
+        )
+      );
+    }
+
+    const verificationTokenEntity = maybeVerificationTokenEntity.value;
+    const {
+      InvalidAfter: invalidAfter,
+      FiscalCode: fiscalCode
+    } = verificationTokenEntity;
+
+    // Check if the token is expired
+    if (Date.now() > new Date(invalidAfter).getTime()) {
```

Comment:
use a decoder to get `invalidAfter` as `Date` - btw why don't we encode the creation date instead? if we do that we can change the duration of existing tokens at any time